File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
public:
bool checkInclusion(string s1, string s2) {
   vector<int> hash1(26,0), hash2(26,0);
   if ( s1.size() > s2.size() ) return false;
   for ( int i = 0; i < s1.size(); i++ )
   {
      hash1[s1[i] - 'a']++;
      hash2[s2[i] - 'a']++;
   }
   for ( int l = 0, r = s1.size() - 1; r < s2.size() - 1;)
   {
      if ( hash2 == hash1 ) return true;
      hash2[s2[l] - 'a']--;
      l++;
      r++;
      hash2[s2[r] - 'a']++;
   }
   return hash2 == hash1;
}
};

File: 0567-permutation-in-string/0567-permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if (len(s1) > len(s2)): return False
        s1_d, s2_d = [0] * 26, [0] * 26

        for s in s1:
            s1_d[ord(s) - ord('a')] += 1

        l = 0
        for r in range(len(s2)):
            if (s2_d == s1_d): return True
            index = ord(s2[r]) - ord('a')
            s2_d[index] += 1
            while (s2_d[index] > s1_d[index]):
                s2_d[ord(s2[l]) - ord('a')] -= 1
                l += 1
            
        return s2_d == s1_d